pkg/virt-controller/watch/replicaset.go

```diff
@@ -0,0 +1,533 @@
+	if err != nil {
+		log.Error().Reason(err).Msg("Invalid selector on replicaset, will not re-enqueue.")
+		return nil
+	}
+
+	needsSync := c.expectations.SatisfiedExpectations(key)
+
+	// get all potentially interesting VMs from the cache
+	vms, err := c.listVMsFromNamespace(rs.ObjectMeta.Namespace)
+
+	if err != nil {
+		log.Error().Reason(err).Msg("Failed to fetch vms for namespace from cache.")
+		return err
+	}
+
+	// make sure we only consider active VMs
+	vms = c.filterActiveVMs(vms)
+
+	// make sure the selector of the controller matches and the VMs match
+	vms = c.filterMatchingVMs(selector, vms)
+
+	// Scale up or down, if all expected creates and deletes were report by the listener
+	var scaleErr error
+	if needsSync {
+		scaleErr = c.scale(rs, vms)
+	}
+
+	if scaleErr != nil {
+		log.Error().Reason(err).Msg("Scaling the replicaset failed.")
+	}
+
+	clone, err := model.Clone(rs)
+
+	if err != nil {
+		log.Error().Reason(err).Msg("Cloning the replicaset failed.")
+		return nil
+	}
+	rsCopy := clone.(*virtv1.VirtualMachineReplicaSet)
+
+	err = c.updateStatus(rsCopy, vms, scaleErr)
+	if err != nil {
+		log.Error().Reason(err).Msg("Updating the replicaset status failed.")
+	}
+
+	return err
+}
+
+func (c *VMReplicaSet) scale(rs *virtv1.VirtualMachineReplicaSet, vms []virtv1.VirtualMachine) error {
+
+	diff := c.calcDiff(rs, vms)
+	rsKey, err := kubecli.KeyFunc(rs)
+	if err != nil {
+		logging.DefaultLogger().Error().Object(rs).Reason(err).Msg("Failed to extract rsKey from replicaset.")
+		return nil
+	}
+
+	if diff == 0 {
+		return nil
+	}
+
+	// Every delete request can fail, give the channel enough room, to not block the go routines
+	errChan := make(chan error, abs(diff))
+
+	var wg sync.WaitGroup
+	wg.Add(abs(diff))
+
+	if diff > 0 {
+		// We have to delete VMs, use a very simple se
+		deleteCandidates := vms[0:diff]
+		c.expectations.ExpectDeletions(rsKey, kubecli.VirtualMachineKeys(deleteCandidates))
+		for i := 0; i < diff; i++ {
+			go func(idx int) {
+				defer wg.Done()
+				deleteCandidate := &vms[idx]
+				// TODO graceful delete
+				err := c.clientset.VM(rs.ObjectMeta.Namespace).Delete(deleteCandidate.ObjectMeta.Name, &v1.DeleteOptions{})
+				// Don't log an error if it is already deleted
+				if err != nil {
+					// We can't observe a delete if it was not accepted by the server
+					c.expectations.DeletionObserved(rsKey, kubecli.VirtualMachineKey(deleteCandidate))
+					c.recorder.Eventf(deleteCandidate, k8score.EventTypeWarning, FailedDeleteVirtualMachineReason, "Error deleting: %v", err)
+					errChan <- err
+					return
+				}
+				c.recorder.Eventf(deleteCandidate, k8score.EventTypeNormal, SuccessfulDeleteVirtualMachineReason, "Deleted virtual machine: %v", deleteCandidate.ObjectMeta.UID)
+			}(i)
+		}
+
+	} else if diff < 0 {
+		// We have to create VMs
+		c.expectations.ExpectCreations(rsKey, abs(diff))
+		basename := c.getVirtualMachineBaseName(rs)
+		for i := diff; i < 0; i++ {
```

Comment:
Something like

```golang
for i := min(diff, burstReplicas); i < 0; i ++ {}
```

for creations and deletions should give us the protection.